### ISSUE TITLE
ESQL searches only case-sensitive

### ISSUE THREAD

**Issue by Mat0vu** [SigmaHQ/pySigma-backend-elasticsearch#107]
Hi everyone,

last week I discovered some issues regarding case-insensitive search in Elasticsearch using `ESQL`.
I was testing something with a very simple rule, trying to find a commandline containing `git.exe`:

```yaml
detection:
  condition: selection
  selection:
    CommandLine|contains: git.exe
```
which resulted in `... | where process.command_line Like "*git.exe*"` (using the ecs pipeline) and was working as expected until I´ve changed the search term for `*Git.exe*`. With this there was no hit using ESQL.

In the ecs documentation I found that the regular `process.command_line` is of type keyword but contains a `.text` subfield which should be case-insensitive. So I did some experiments with both fields and using ESQL in comparison to KQL.

| KQL Search Query | ESQL Search Query | Found with KQL | Found with ESQL | Comment |
|------------------------------|---------------------------------|-----------------|-----------------|----|
| process.command_line: *git.exe* | Where process.command_line Like "*git.exe*" | Yes | Yes | |  
| process.command_line: *Git.exe* | Where process.command_line Like "*Git.exe*" | No | No | | 
| process.command_line.text: *git.exe* | Where process.command_line.text Like "*git.exe*" | Yes | Yes | | 
| process.command_line.text: *Git.exe* | Where process.command_line.text Like "*Git.exe*" | Yes | No |  `.text` subfield is case-**in**sensitive for `KQL`; `ESQL` is still case-sensitive |

I was expecting that when I use the `.text` subfield I would also find something with ESQL, because the field should allow for case-insensitive searches but this was not the case.
After some digging in the github repo of Elastic I found out that they had case-insensitive operators implemented for ESQL but they were disabled again and there doesn´t seem to be any current developments (https://github.com/elastic/elasticsearch/issues/105603).

We got in contact with the support to check how we can get a case-insensitive search behaviour using ESQL and got the feedback that:
- `like` and `rlike` operators are case-sensitive in `ESQL`
- to achieve case-insenitive behaviour we should use the `TO_LOWER` or `TO_UPPER` function in combination with a lower-case/upper-case search term, e.g. to match `IntelliJ` one could use `... | Where TO_LOWER(process.command_line) Like "*intellij*" `

I tested this approach and it is working although I think it is not very nice to need such a function.

To sum it up, currently the `ESQL` backend is searching case-sensitive which is not following the sigma specification to treat all strings as case-insensitive and it will lead to missed detections. 
Probably the only way to get case-insensitive behaviour is by following the suggestion from the support and implement these functions in the backend. What do you think?

**Comment by abulhol** [comment]
Please note I have incurred another big problem with LIKE operators in ESQL:
https://github.com/elastic/elasticsearch/issues/118932
This lets a total of 645 rules (~25% of all rules) from the main Sigma repo fail. 

We definitely need to do some work on this project here to make Sigma fully work with ESQL.

**Comment by thomaspatzke** [comment]
Oh my...Elastic case-sensitiveness really drives me crazy 🙁

Just to clarify, this:

> * `like` and `rlike` operators are case-sensitive in `ESQL`

Is falsified by your insights, correct?

`TO_LOWER` seems the only option to me here, but I fear this could result in a really bad query performance. Did you tests of the performance?

**Comment by Mat0vu** [comment]
Hi Thomas,

unfortunately the statement 
> `like` and `rlike` operators are **case-sensitive** in `ESQL`

is true and also confirmed my results from the little experiment.
Unless these operators are combined with something like `TO_LOWER`, they will miss all strings with a single different capitalization.

The only case-**in**sensitive operator that I have found in `ESQL` is `=~`. However, this operator is currently deactivated again (see the linked issue in the initial text) and **would only return full matches** anyways.

Since the Elastic-support wrote that `(r)like` +  `TO_LOWER` is the only solution, I fear that right now and also for the near future there is no alternative way, however I havent tested the performance impact yet. I can try to run some experiments in the new year and also ask Elastic if they can give out some statistics about this. If you agree that we should follow this path, I can also start adapting the `ESQL`-Backend to use `TO_LOWER`...

I agree with you, this topic is a real pain :( 
Also, I do not understand why `ESQL` does not support `(r)like~`, which works case-**in**sensitive, but only for `EQL`...

Anyways, happy new year in advance :)

**Comment by Mat0vu** [comment]
Hi,

I wanted to share the latest information from the Elastic Support regarding ESQL:

- There are currently no plans to make ESQL case-insensitive by default.
- The only method to perform case-insensitive searches in ESQL is by using the TO_UPPER/TO_LOWER functions (see https://github.com/elastic/elasticsearch/issues/118304).
- While they are not able/willing to share specific benchmarks, some customers have reported significant performance decreases to them. I won't disclose the exact figures due to confidentiality statements in the support communications, but the performance impact is quite substantial, particularly when multiple functions are used on various fields within a single query.

Given the lack of alternatives, I have begun updating the ESQL backend to incorporate these functions (you can find the progress here: https://github.com/Mat0vu/pySigma-backend-elasticsearch/tree/case-insensitive-esql). This still requires some additional work and will contain a lot of changes because all unit tests must also be revised to accommodate the need for lowercase search strings.

It is probably needed to include a warning indicating that ESQL may not be the best option for basic search operations and that users are encouraged to consider alternative languages when possible. Especially if the rule does not exclusively utilize the `|cased` sigma operators, which likely applies for most rules.

**Comment by andurin** [comment]
@Mat0vu 

Maybe helpful or not. I'm working on another backend and had the same issue regarding case-insensitive searches.

The system also only have a "to_lower()" function and I had to change both, the field name and also the value to have the to_lower() on both sides.

I helped myself by first having a processing pipeline to suffix the field name:

```
transformations:
  - id: custom_field_suffix_tolower
    type: field_name_suffix
    suffix: ".to_lower()"
    field_name_conditions:
      - type: include_fields
        fields:
          - process.cmd_line
          - process.parent_process.cmd_line
          - process.name
          - process.parent_process.name
```

and after that taking a conditional change in the backend to also suffix the value:
```
    def convert_condition_field_eq_val_str(
        self, cond: ConditionFieldEqualsValueExpression, state: ConversionState
    ) -> Union[str, DeferredQueryExpression]:
        """Conversion of field = string value expressions"""
        if ".to_lower()" in cond.field:
            subitem = self.convert_value_str(cond.value[1:-1], state)
            return (
                super()
                .convert_condition_field_eq_val_str(cond, state)
                .replace(subitem, f"{subitem}.to_lower()")
            )
        else:
            return super().convert_condition_field_eq_val_str(cond, state)
```

**Comment by Mat0vu** [comment]
Hi @andurin ,

thanks for the tip, at first glance that looks quite helpful. For the specific case of Elastic I´ve experimented with the following pipeline, to get the necessary output `TO_LOWER(field_name)=="value"`:

```yaml
transformations:
  - id: custom_field_prefix_tolower
    type: field_name_prefix
    prefix: TO_LOWER(
  - id: custom_field_suffix_tolower
    type: field_name_suffix
    suffix: )
```

However I think that there is a special case for which this approach won´t produce the correct ouput: When the parameter `|cased` is used, then no `TO_LOWER` should be added as the match should be case-sensitive. 
I think the use of `|cased` cannot be detected by the pipeline using the available conditions or am I wrong here? Because if this would be possible, then this approach would be very elegant.

Until now I´ve worked on a not so elegant solution to this issue by adjusting/overwriting the code of the (ESQL) backend code (especially the methods `convert_condition_field_eq_val_str`, `convert_condition_field_eq_val_str_case_sensitive` and `convert_condition_as_in_expression`, see https://github.com/Mat0vu/pySigma-backend-elasticsearch/blob/case-insensitive-esql/sigma/backends/elasticsearch/elasticsearch_esql.py#L590 and following). 
As far as I´ve tested it, this approach is working as expected and also pushes down the values used to search. For that also the `convert_value_str` method is overwritten, adding a parameter `enforce_lowercase` which is set to `True` when an `TO_LOWER` is added but stays `False` for sensitive matches. By handling this in the distinct methods for case-(in)sensitive matching, it´s also correctly handling the `|cased` use-case. 
I´ve also updated/added some unit tests on this branch but have not yet created a PR because it is based on a newer pysigma version...

Clearly my current approach is not as nice as a simple pipeline because quite a lot of code is changed but I´m not sure if it can be fully addressed using pipelines. What do you think?
